File: 113.path-sum-ii.py
#
# @lc app=leetcode id=113 lang=python3
#
# [113] Path Sum II
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        res = []
        self.dfs(root, targetSum, 0, res, [])
        return res

    def dfs(self, root, targetSum, path_sum, res, path: list[int]):
        if not root:
            return
        
        path_sum += root.val
        path.append(root.val)
    
        if path_sum == targetSum and not root.left and not root.right:
            res.append(list(path))
        

        self.dfs(root.left, targetSum, path_sum, res, path)
        self.dfs(root.right, targetSum, path_sum, res, path)

        path.pop()

        
# @lc code=end

File: 113.path-sum-ii.java
/*
 * @lc app=leetcode id=113 lang=java
 *
 * [113] Path Sum II
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        return this.dfs(root, targetSum, 0, new ArrayList<>(), new ArrayList<>());
    }

    public List<List<Integer>> dfs(TreeNode root, int targetSum, int pathSum, List<List<Integer>> res, List<Integer> path) {
        
        if (root == null) {
            return res;
        }

        path.add(root.val);
        pathSum += root.val;

        if (pathSum == targetSum && root.left == null && root.right == null) {
            res.add(new ArrayList<>(path));
        }

        this.dfs(root.left, targetSum, pathSum, res, path);
        this.dfs(root.right, targetSum, pathSum, res, path);

        path.remove(path.size() - 1);
        
        return res;
    }
}
// @lc code=end

